stats/patient_interceptor.go

```diff
@@ -0,0 +1,127 @@
+package stats
+
+import (
+	"errors"
+	"log"
+
+	"github.com/intervention-engine/fhir/models"
+)
+
+// PatientStatsCreateInterceptor intercepts any new patient resources added to the database
+// and updates the Synthetic Mass population statistics based on that patient's address.
+type PatientStatsCreateInterceptor struct {
+	DataAccess StatsDataAccess
+}
+
+// NewPatientStatsCreateInterceptor returns an initialized instance of PatientStatsCreateInterceptor
+func NewPatientStatsCreateInterceptor(pgDataAccess StatsDataAccess) *PatientStatsCreateInterceptor {
+	return &PatientStatsCreateInterceptor{DataAccess: pgDataAccess}
+}
+
+// Before is unused
+func (s *PatientStatsCreateInterceptor) Before(resource interface{}) {}
+
+// After increments population statistics after a patient resource is created.
+func (s *PatientStatsCreateInterceptor) After(resource interface{}) {
+	patient, ok := resource.(*models.Patient)
+
+	if ok {
+		err := s.DataAccess.AddPatientStat(patient)
+		if err != nil {
+			log.Printf("PatientStatsCreateInterceptor: After: %s\n", err.Error())
+		}
+	}
+}
+
+// OnError is unused
+func (s *PatientStatsCreateInterceptor) OnError(err error, resource interface{}) {}
+
+// PatientStatsUpdateInterceptor intercepts any updated patient resources
+// and updates the Synthetic Mass population statistics based on that patient's address.
+type PatientStatsUpdateInterceptor struct {
+	DataAccess StatsDataAccess
+	// The state of the patient before the database update, for comparison after the database update
+	patientsBefore map[string]*models.Patient
+	// Tracks if the interceptor failed to cache the patient model before the update
+	cacheError error
```

Comment:
Please see my comments on the ConditionInterceptor (in regards to `cacheError` and when to delete the entry from the map.  They apply here too.